File: app/src/main/java/jp/hazuki/yuzubrowser/utils/view/swipebutton/SwipeSoftButtonController.java
package jp.hazuki.yuzubrowser.utils.view.swipebutton;

import android.content.Context;
import android.graphics.drawable.Drawable;

import jp.hazuki.yuzubrowser.action.ActionCallback;
import jp.hazuki.yuzubrowser.action.manager.SoftButtonActionFile;

public class SwipeSoftButtonController extends SwipeController {
    private ActionCallback mCallback = null;
    private SoftButtonActionFile mActionList = null;
    //private Drawable mBackgroundDrawable;

    public SwipeSoftButtonController(Context context) {
        super(context);
    }

    public void setActionData(SoftButtonActionFile actionlist, ActionCallback callback) {
        mActionList = actionlist;
        mCallback = callback;
    }

    public Drawable getIcon() {
        return getIcon(getCurrentWhatNo());
    }

    public Drawable getIcon(int whatNo) {
        if (mCallback == null) return null;
        switch (whatNo) {
            case SWIPE_CANCEL:
            case SWIPE_PRESS:
                return mCallback.getIcon(mActionList.press);
            case SWIPE_LPRESS:
                return mCallback.getIcon(mActionList.lpress);
            case SWIPE_UP:
                return mCallback.getIcon(mActionList.up);
            case SWIPE_DOWN:
                return mCallback.getIcon(mActionList.down);
            case SWIPE_LEFT:
                return mCallback.getIcon(mActionList.left);
            case SWIPE_RIGHT:
                return mCallback.getIcon(mActionList.right);
        }
        return null;
    }

    public Drawable getDefaultIcon() {
        if (mCallback == null) return null;
        return mCallback.getIcon(mActionList.press);
    }

    @Override
    protected void onEventActionUp(int whatNo) {
        //mBackgroundDrawable.setState(STATE_NOTHING);
        if (mActionList != null && mCallback != null) {
            switch (whatNo) {
                case SWIPE_PRESS:
                    mCallback.run(mActionList.press);
                    break;
                case SWIPE_LPRESS:
                    //mCallback.run(mActionList.lpress);
                    break;
                case SWIPE_UP:
                    mCallback.run(mActionList.up);
                    break;
                case SWIPE_DOWN:
                    mCallback.run(mActionList.down);
                    break;
                case SWIPE_LEFT:
                    mCallback.run(mActionList.left);
                    break;
                case SWIPE_RIGHT:
                    mCallback.run(mActionList.right);
                    break;
            }
        }
    }

    protected void onEventActionDown() {
        //mBackgroundDrawable.setState(STATE_PRESSED);
    }

    protected void onEventCancel() {
        //mBackgroundDrawable.setState(STATE_NOTHING);
    }

    protected void onEventOutSide() {
        //mBackgroundDrawable.setState(STATE_NOTHING);
    }

    @Override
    protected void onEventLongPress() {
        if (mActionList != null && mCallback != null) {
            mCallback.run(mActionList.lpress);
        }
    }

    public boolean shouldShow() {
        return !mActionList.press.isEmpty();
    }

    /*public void setBackgroundDrawable(Drawable background) {
        mBackgroundDrawable = background;
	}
	
	private static final int[] STATE_NOTHING = new int[] {0};
	private static final int[] STATE_PRESSED = new int[] {android.R.attr.state_pressed};
	
	public void getBackgroundDrawable(Drawable background) {
		mBackgroundDrawable = background;
		mBackgroundDrawable.setState(STATE_NOTHING);
	}
	
	public Drawable getCurrentBackground() {
		return mBackgroundDrawable.getCurrent();
	}*/
}

File: app/src/main/java/jp/hazuki/yuzubrowser/utils/view/swipebutton/SwipeSoftButtonController.kt
/*
 * Copyright (C) 2017-2018 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.utils.view.swipebutton

import android.content.Context
import android.graphics.drawable.Drawable
import jp.hazuki.yuzubrowser.action.manager.ActionController
import jp.hazuki.yuzubrowser.action.manager.ActionIconManager
import jp.hazuki.yuzubrowser.action.manager.SoftButtonActionFile

class SwipeSoftButtonController
//private Drawable mBackgroundDrawable;

(context: Context) : SwipeController(context) {
    private var mActionList: SoftButtonActionFile? = null
    private var controller: ActionController? = null
    private var iconManager: ActionIconManager? = null

    val icon: Drawable?
        get() = getIcon(currentWhatNo)

    val defaultIcon: Drawable?
        get() {
            val actionList = mActionList ?: return null
            val iconManager = iconManager ?: return null

            return iconManager[actionList.press]
        }

    fun setActionData(actionlist: SoftButtonActionFile, controller: ActionController, iconManager: ActionIconManager) {
        mActionList = actionlist

        this.controller = controller
        this.iconManager = iconManager
    }

    fun getIcon(whatNo: Int): Drawable? {
        val actionList = mActionList ?: return null
        return iconManager?.let {
            when (whatNo) {
                SwipeController.SWIPE_CANCEL, SwipeController.SWIPE_PRESS -> it[actionList.press]
                SwipeController.SWIPE_LPRESS -> it[actionList.lpress]
                SwipeController.SWIPE_UP -> it[actionList.up]
                SwipeController.SWIPE_DOWN -> it[actionList.down]
                SwipeController.SWIPE_LEFT -> it[actionList.left]
                SwipeController.SWIPE_RIGHT -> it[actionList.right]
                else -> null
            }
        }
    }

    override fun onEventActionUp(whatNo: Int) {
        //mBackgroundDrawable.setState(STATE_NOTHING);
        val actionList = mActionList ?: return
        val controller = controller ?: return

        when (whatNo) {
            SwipeController.SWIPE_PRESS -> controller.run(actionList.press)
        //SwipeController.SWIPE_LPRESS -> Nothing
            SwipeController.SWIPE_UP -> controller.run(actionList.up)
            SwipeController.SWIPE_DOWN -> controller.run(actionList.down)
            SwipeController.SWIPE_LEFT -> controller.run(actionList.left)
            SwipeController.SWIPE_RIGHT -> controller.run(actionList.right)
        }
    }

    override fun onEventActionDown() {
        //mBackgroundDrawable.setState(STATE_PRESSED);
    }

    override fun onEventCancel() {
        //mBackgroundDrawable.setState(STATE_NOTHING);
    }

    override fun onEventOutSide() {
        //mBackgroundDrawable.setState(STATE_NOTHING);
    }

    override fun onEventLongPress() {
        if (mActionList != null && controller != null) {
            controller!!.run(mActionList!!.lpress)
        }
    }

    fun shouldShow(): Boolean {
        return !mActionList!!.press.isEmpty()
    }
}
